### PR TITLE
Bugfix for MNASNet

### PR DESCRIPTION
**Pull request by 1e100**
The original implementation I submitted contained a bug which affects all MNASNet variants other than 1.0. The bug is that the first few layers need to also be scaled in terms of width multiplier, along with all the rest. This fixes the issue, and brings the implementation fully in sync with Google's TPU reference code. I have compared the ONNX dump of this model against TFLite's hosted model and ensured that all layer configurations line up exactly.

Because only MNASNet 0.5 checkpoint was affected, I have also trained a slightly better checkpoint for it. I was unable to train this to the same accuracy with Torchvision's reference training code (and it wasn't for the lack of trying), and had to use label smoothing and EMA to get this result. The final checkpoint is derived from EMA. 

Even so, the accuracy is a bit lower than Google's result (67.83 for this model vs 68.03 TFLite). The posted number for TF TPU implementation is 68.9, but that model uses SE on some of its layers, which my implementation does not.

### CODE REVIEW

**Comment by 1e100** [comment]
Sure, we can make it backward compatible. Of the two proposed solutions I like solution #1 better (fewer moving parts), but I'll implement whatever we decide here. 

Maybe I should take this opportunity and add squeeze and excitation as well. Authors use it in some of the larger variants of the model. Accuracy will go up a bit if I do that, and then we won't have to version it again in the future.

Is there a release schedule BTW? I wanted this fix to be ready before the release (models unfortunately take forever to train), and noticed the release had been cut a few days ago, so the PR didn't make it.

**Comment by fmassa** [comment]
Hi @1e100 ,

I just got back from holidays, sorry for the delay in replying.

> Sure, we can make it backward compatible. Of the two proposed solutions I like solution 1 better (fewer moving parts), but I'll implement whatever we decide here.

I'm not yet clear on the best solution. I feel that this is something that needs to be carefully considered, because model versioning is going to be a big topic.

> Maybe I should take this opportunity and add squeeze and excitation as well.

I feel that this should be sent in a separate PR.

> Is there a release schedule BTW? 

We will be cutting a new release of torchvision in the next 2-3 weeks, with minor fixes and improvements.


I'm also tagging @ailzhang for handling BC-breaking changes within hub, and @cpuhrsch @vincentqb and @zhangguanheng66 for torchaudio and torchtext model versioning in the future.

**Comment by soumith** [comment]
using `_version` counter like BatchNorm makes a lot of sense to me.

**Comment by 1e100** [comment]
OK, had some time today to look into this, aiming to get it done over the weekend. 

Basically, it seems that the following simple logic would satisfy the backward compat requirements:

 1. Add `_version == 1` field
 2. When model is loaded using load_state_dict(), if `_version` field is not found in the dict, modify the network definition such that the stem is configured as it was configured in the initial version and perhaps show a warning as in `MultiHeadAttention`, as well as remove the `_version` field, turning the model into a de-facto previous version model. If `_version == 1` is found, do nothing and simply load the new checkpoint.

Seems like a pretty straightforward fix to me.

**Comment by 1e100** [comment]
OK, @fmassa, here's the first cut of the requested changes. CI seems to be failing on something CUDA-related. Let me know if this is what you had in mind!

File: torchvision/models/mnasnet.py [hunk 1]
```diff
@@ -90,30 +93,37 @@ class MNASNet(torch.nn.Module):
...
     """
+    # Version 2 adds depth scaling in the initial stages of the network.
+    _version = 2
 
     def __init__(self, alpha, num_classes=1000, dropout=0.2):
         super(MNASNet, self).__init__()
-        depths = _scale_depths([24, 40, 80, 96, 192, 320], alpha)
+        assert alpha > 0.0
+        self.alpha = alpha
+        self.variant = "b1"
```
[the remark this answers is not in the document]
Comment (1e100) [reply torchvision/models/mnasnet.py]: At this point this is just for identifying the model variant explicitly. If support for other model types is added, this variable would refer to the variant, which is useful when reading from a checkpoint.

File: torchvision/models/mnasnet.py [hunk 2]
```diff
@@ -8,7 +9,7 @@
 
 _MODEL_URLS = {
     "mnasnet0_5":
-    "https://download.pytorch.org/models/mnasnet0.5_top1_67.592-7c6cb539b9.pth",
+    "https://github.com/1e100/mnasnet_trainer/releases/download/0.3/mnasnet0.5_top1_67.823-3ffadce67e.pth",
```
[the remark this answers is not in the document]
Comment (1e100) [reply torchvision/models/mnasnet.py]: Could you elaborate? Older code will still be able to read the older checkpoint, and it's perfectly OK if you keep serving it. Newer code should read the new checkpoint. The checkpoints are not compatible across model versions, so if new code reads an older checkpoint, the model downgrades itself to v1. This is justifiable when you already have a transfer-learned model, but IMO undesirable in case of a pretrained checkpoint.

File: torchvision/models/mnasnet.py [hunk 3]
```diff
@@ -139,16 +149,58 @@ def _initialize_weights(self):
                 nn.init.ones_(m.weight)
                 nn.init.zeros_(m.bias)
             elif isinstance(m, nn.Linear):
-                nn.init.normal_(m.weight, 0.01)
+                nn.init.kaiming_uniform_(m.weight, mode="fan_out",
```
[the remark this answers is not in the document]
Comment (1e100) [reply torchvision/models/mnasnet.py]: It may have very slightly improved the top1 on MNASNet b1 0.5 that I trained for this PR, but I'm not sure. The purpose of the change is that initialization is now identical to the reference TensorFlow code (which also uses a variance scaling initializer aka Kaiming uniform). Certainly not worse than before.

File: torchvision/models/mnasnet.py [hunk 4]
```diff
@@ -139,16 +149,58 @@ def _initialize_weights(self):
...
+                nn.ReLU(inplace=True),
+                nn.Conv2d(32, 16, 1, padding=0, stride=1, bias=False),
+                nn.BatchNorm2d(16, momentum=_BN_MOMENTUM),
+                _stack(16, depths[2], 3, 2, 3, 3, _BN_MOMENTUM),
+            ]
+            for idx, layer in enumerate(v1_stem):
+                self.layers[idx] = layer
+
+            # The model is now identical to v1, and must be saved as such.
+            MNASNet._version = 1
```
[the remark this answers is not in the document]
Comment (1e100) [reply torchvision/models/mnasnet.py]: D'oh! You're right. Changed, and verified it works with this code:
```python3
#!/usr/bin/env python3

import torch
import torchvision

ckpt = torch.load("mnasnet0.5_top1_67.823-3ffadce67e.pth")
m = torchvision.models.MNASNet(0.5)
m.load_state_dict(ckpt)
print("Loaded old")
torch.save(m.state_dict(), "mnasnet0.5_top1_67.823.pth")
print("Re-saved")
ckpt = torch.load("mnasnet0.5_top1_67.823.pth")
m = torchvision.models.MNASNet(0.5)
m.load_state_dict(ckpt)
print("Re-loaded")
```

**Comment by 1e100** [comment]
@fmassa I've addressed your feedback, PTAL

File: torchvision/models/mnasnet.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fmassa) [reply torchvision/models/mnasnet.py]: I'd say that it's then better to remove this for now, and add it back once (if) we decide to add other variants

File: torchvision/models/mnasnet.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fmassa) [reply torchvision/models/mnasnet.py]: I'm not sure if I agree.

If the user update their version of torchvision, their pre-trained classifier (based on stock MNasNet features from torchvision) will not work anymore, and this is definitely not expected.

**Comment by fmassa** [comment]
Sorry for the delay in replying, I've made a few more comments

File: torchvision/models/mnasnet.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (1e100) [reply torchvision/models/mnasnet.py]: Not sure I understand. If they load the new weights provided here it will work fine. If they load their old transfer-learned model checkpoint, it will also work fine because the model turns into a v1 model onload. What's the scenario we're trying to guard agains here?

File: torchvision/models/mnasnet.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (1e100) [reply torchvision/models/mnasnet.py]: Done!

File: torchvision/models/mnasnet.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fmassa) [reply torchvision/models/mnasnet.py]: If the users only saved the classifier, and not the backbone (which was fixed).
Here is an example:
```python
model = torchvision.models.mnasnet(pretrained=True)
features = model.features
classifier = nn.Linear(2, 2)
# train classifier, not backproping through the feature
# extractor to save compute
with torch.no_grad():
    f = features(input)
output = classifier(f)
loss = F.mse(output, target)
loss.backward()

# and then, save classifier only
torch.save(classifier, 'file.pth')
```
I think this can be a reasonable scenario. And this means that they will need the pre-trained weights for the old models as well.

[the remark this answers is not in the document]
Comment (fmassa) [reply torchvision/models/mnasnet.py]: After discussion with @soumith , we decided that we will not take this case into account.
We should just hard-error and replace the old parameters with the new ones.

So I think the current status of the PR is good, thanks!

**Comment by fmassa** [comment]
@1e100 I couldn't push the updated path without a force push on your master branch.

Can you update the link in the PR to
```
https://download.pytorch.org/models/mnasnet0.5_top1_67.823-3ffadce67e.pth
```
And let me know?

**Comment by 1e100** [comment]
@fmassa done!

**Comment by fmassa** [comment]
Thanks a lot!
